File: health_new_p04/backend/models/health_model.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum

from pydantic import BaseModel, Field, field_validator


class IngestionSource(str, Enum):
    BLE = "ble"
    MQTT = "mqtt"
    SERIAL = "serial"
    MOCK = "mock"


class HealthSample(BaseModel):
    device_mac: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    heart_rate: int = Field(ge=0, le=240)
    temperature: float = Field(ge=0.0, le=45.0)
    blood_oxygen: int = Field(ge=0, le=100)
    blood_pressure: str | None = None
    battery: int = Field(default=0, ge=0, le=100)
    sos_flag: bool = False
    source: IngestionSource = IngestionSource.MOCK
    device_uuid: str | None = None
    ambient_temperature: float | None = None
    surface_temperature: float | None = None
    steps: int | None = None
    packet_type: str | None = None
    sos_value: int | None = None
    sos_trigger: str | None = None
    raw_packet_a: str | None = None
    raw_packet_b: str | None = None
    anomaly_score: float | None = None
    health_score: int | None = None

    @field_validator("device_mac")
    @classmethod
    def normalize_mac(cls, value: str) -> str:
        return value.upper()

    @property
    def blood_pressure_pair(self) -> tuple[int, int]:
        if not self.blood_pressure:
            return 120, 80
        try:
            systolic, diastolic = self.blood_pressure.split("/", maxsplit=1)
            return int(systolic), int(diastolic)
        except (TypeError, ValueError):
            return 120, 80


class HealthTrendPoint(BaseModel):
    timestamp: datetime
    heart_rate: int
    temperature: float
    blood_oxygen: int
    blood_pressure: str | None = None
    steps: int | None = None
    health_score: int | None = None


class IngestResponse(BaseModel):
    success: bool = True
    message: str = "Sample ingested"
    device_mac: str | None = None
    sample: HealthSample | None = None
    triggered_alarm_ids: list[str] = Field(default_factory=list)

File: health_new_p04/mobile/flutter_app/lib/features/health/models/health_model.dart
class HealthData {
  final String deviceMac;
  final DateTime timestamp;
  final double? heartRate;
  final double? temperature;
  final double? bloodOxygen;
  final String? bloodPressure;
  final int? battery;
  final bool sosFlag;
  final int? steps;
  final int? healthScore;

  HealthData({
    required this.deviceMac,
    required this.timestamp,
    this.heartRate,
    this.temperature,
    this.bloodOxygen,
    this.bloodPressure,
    this.battery,
    this.sosFlag = false,
    this.steps,
    this.healthScore,
  });

  factory HealthData.fromJson(Map<String, dynamic> json) {
    return HealthData(
      deviceMac: json['device_mac'] as String? ?? '',
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'] as String)
          : DateTime.now(),
      heartRate: (json['heart_rate'] as num?)?.toDouble(),
      temperature: (json['temperature'] as num?)?.toDouble(),
      bloodOxygen: (json['blood_oxygen'] as num?)?.toDouble(),
      bloodPressure: json['blood_pressure'] as String?,
      battery: json['battery'] as int?,
      sosFlag: json['sos_flag'] == true || (json['sos_value'] as num? ?? 0) > 0,
      steps: json['steps'] as int?,
      healthScore: json['health_score'] as int?,
    );
  }

  HealthData mergeWith(HealthData incoming) {
    return HealthData(
      deviceMac: incoming.deviceMac.isNotEmpty ? incoming.deviceMac : deviceMac,
      timestamp: incoming.timestamp,
      heartRate: _preferPositiveNumber(heartRate, incoming.heartRate),
      temperature: _preferPositiveNumber(temperature, incoming.temperature),
      bloodOxygen: _preferPositiveNumber(bloodOxygen, incoming.bloodOxygen),
      bloodPressure:
          _preferBloodPressure(bloodPressure, incoming.bloodPressure),
      battery: _preferIncrementalInt(battery, incoming.battery),
      sosFlag: incoming.sosFlag,
      steps: _preferIncrementalInt(steps, incoming.steps),
      healthScore: _preferIncrementalInt(healthScore, incoming.healthScore),
    );
  }

  static double? _preferPositiveNumber(double? previous, double? incoming) {
    if (incoming != null && incoming > 0) return incoming;
    return previous ?? incoming;
  }

  static int? _preferIncrementalInt(int? previous, int? incoming) {
    if (incoming != null && incoming > 0) return incoming;
    return previous ?? incoming;
  }

  static String? _preferBloodPressure(String? previous, String? incoming) {
    return _normalizeBloodPressure(incoming) ??
        _normalizeBloodPressure(previous);
  }

  static String? _normalizeBloodPressure(String? value) {
    if (value == null || value.trim().isEmpty) return null;
    final parts = value.split('/');
    if (parts.length != 2) return null;
    final systolic = int.tryParse(parts[0].trim());
    final diastolic = int.tryParse(parts[1].trim());
    if (systolic == null || diastolic == null) return null;
    if (systolic <= 0 || diastolic <= 0) return null;
    return '$systolic/$diastolic';
  }
}
